x/auth/client/cli/tx_sign.go

```diff
@@ -230,58 +231,49 @@ func makeSignCmd() func(cmd *cobra.Command, args []string) error {
 			return err
 		}
 
-		// if --signature-only is on, then override --append
-		generateSignatureOnly, _ := cmd.Flags().GetBool(flagSigOnly)
-		multisigAddrStr, _ := cmd.Flags().GetString(flagMultisig)
-
-		from, _ := cmd.Flags().GetString(flags.FlagFrom)
+		generateSignatureOnly, _ := f.GetBool(flagSigOnly)
+		multisigAddrStr, _ := f.GetString(flagMultisig)
+		from, _ := f.GetString(flags.FlagFrom)
 		_, fromName, err := client.GetFromFields(txFactory.Keybase(), from, clientCtx.GenerateOnly)
 		if err != nil {
 			return fmt.Errorf("error getting account from keybase: %w", err)
 		}
 
 		if multisigAddrStr != "" {
 			var multisigAddr sdk.AccAddress
-
 			multisigAddr, err = sdk.AccAddressFromBech32(multisigAddrStr)
 			if err != nil {
 				return err
 			}
-
 			err = authclient.SignTxWithSignerAddress(
 				txF, clientCtx, multisigAddr, fromName, txBuilder, clientCtx.Offline,
 			)
 			generateSignatureOnly = true
 		} else {
-			flagAppend, _ := cmd.Flags().GetBool(flagAppend)
-			appendSig := flagAppend && !generateSignatureOnly
-			if appendSig {
+			flagAppend, _ := f.GetBool(flagAppend)
+			if flagAppend || generateSignatureOnly {
```

Comment:
This looks incorrect -> `A && !B` != `A || B`

Reply:
the original one was not correct - that was the source of the bug. 
If users sets `generateSignatureOnly` then we should show him a signature.

Reply:
The original one was meant to append a signature if and only if the user didn't select `--signature-only` as well.
Maybe I'm missing something here, but it seems to me that with this change same behaviour is applied if either flags is on

Reply:
I think @alessio is correct here:
- `--sig-only --append` should ignore `--append` and only show sig
- `--sig-only` should only show sign
- `--append` should add a _new_ signature to the tx.

So the only time where we need to call SignTx is when `--append && !--sig-only`.

Reply:
> So the only time where we need to call SignTx is when --append && !--sig-only.

This is not what the user expects - and this was why it was marked as a bug.`--append` is on by default, so only adding `--sig-only` to the CMD will show NO signature if we use that condition (this is what the user did and he is complaining that he doesn't see the signature).

So, putting things together, with `--sig-only` or `--append=false`, as @clevinson  noted, we should overwrite the signature.

Reply:
Then the user has wrong expectations, thus the problem may be the documentation - not the behaviour. The `--signature-only` flag's documentation states clearly:

```
Print only the generated signature, then exit
```

> only adding --sig-only to the CMD will show NO signature if we use that condition

AFAICT this seems to work in Launchpad. Something else broke it.

Reply:
> so only adding --sig-only to the CMD will show NO signature if we use that condition (this is what the user did and he is complaining that he doesn't see the signature).

It will just show the signatures that are already present in the file. If there are no signatures present in that file, then it's normal to show no signature.

Maybe the most intuitive way would be to decouple  `--sig-only` and `--append`:
- `--append` is just a flag denoting if we add a new signature to the tx, or if we overwrite existing signature
- `--sig-only` is just about the final output: whole tx or sig only.

so `if flagAppend { authclient.SignTx() }`, and that's all. This means that `--sig-only --append` will NOT ignore the 2nd flag, but append a new signature and show all signatures.

Reply:
> It will just show the signatures that are already present in the file. If there are no signatures present in that file, then it's normal to show no signature.

I don't think this is an expected behavior. Look at the comment Cory copied from `tx.go` and the flag description: _Append the signature to the existing ones. If disabled, old signatures would be overwritten_

Reply:
> AFAICT this seems to work in Launchpad. Something else broke it.

@alessio  - see my comment above. For me it's not working correctly in Launchpad.

Reply:
> Exactly. This is how I think about it as well.

in this case:

```suggestion
			if flagAppend {
```

Reply:
I did this work long time ago. So yes, this is unarguably the expected behaviour. sigOnly was introduced a a tool for the user to _just_ print the signature. Simple as that.

 If the original intended behaviour is broken, then we should fix it. If the documentation is unclear or ambiguous, then it needs to be improved.

Hope this clarifies.

Reply:
So, @alessio  do you agree that we should override when setting `--flagAppend=false`?

@amaurymartiny , [this](https://github.com/cosmos/cosmos-sdk/pull/8106#discussion_r538303500) is not fully correct. If we agree on override, then we will need to reset the signatures and then add a new signature. ( will update it once we agree on the behavior).

Reply:
The original code may appear a bit convoluted and lack clarity - that's my fault. However, that's how the `sign` command used to work. The original behaviour that IMO should be preserved:

1. `--append=true` just changes the Tx signing behaviour and requires `--signature-only=false`
2. `--signature-only` can be `true` if set so by the user.
3. `--signature-only` should be overriden and set to `true` if a multisig address is passed via the `--multisig` flag too.

In either of those last two cases (2 and 3)  `--append` is overriden and set to `false`.

Reply:
This is the original PR: https://github.com/cosmos/cosmos-sdk/pull/2558

Reply:
@robert-zaremba @amaurymartiny I seem to remember that there used to be test cases for these flags. @robert-zaremba mind writing a test for each case please?